design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,424 @@
+routine may be trying to update this value in parallel with CephCluster reconciliation. If there is
+no parallelism, the internal retry can (should) be removed from the design.
+
+We have the option to target CSI code changes for only the Ceph-CSI Operator. This would allow us to
+have a smaller scope of development, and because we are expecting the CSI Operator to be the primary
+way CSI daemons are managed in the near future, few users would be hindered by the lack of "legacy"
+support.
+
+When `ROOK_ROTATE_CSI_CEPHX_KEYS_OLDER_THAN` is set to a timestamp, Rook should reconcile all
+CephClusters so that keys are updated. ??? Also RbdMirrors and CephFSMirrors ???
+
+### CephRbdMirror example
+
+1. CephRbdMirror keys have both local and non-local keys. For local key rotation, CephCluster's
+    config is used. For the non-local (peer) key, reconciliation reads
+    `spec.security.cephx.rotateKeysOlderThan` to determine if the reconcile should rotate.
+2. Keys will be rotated based on `DATETIME` or the `CurrentCephVersion` string.
+3. Assuming keys are rotated, and successful, the status is updated as so:
+
+    ```yaml
+    status:
+      # ...
+      cephx:
+        local: # report local key rotation status
+          keyReconciliationTimestamp: "Thu May 29 20:18:04 UTC 2025" # e.g.
+          keyCephVersion: "20.2.0" # e.g.
+        peer: # report non-local peer key rotation status
+          keyReconciliationTimestamp: "Fri May 30 18:16:31 UTC 2025" # e.g.
+          keyCephVersion: "20.2.0" # e.g.
+    ```
+
+### ALTERNATIVE FOR DISCUSSION
+
+Instead of `DATETIME` based one-time rotation, we could use "generation" based one-time rotation.
+
+Example, user is initiating local key rotation:
+
+```yaml
+kind: CephCluster
+# ...
+spec:
+  # ...
+  security:
+    cephx:
+      rotateLocalKeysOlderThan: gen2 # e.g., user has just updated this from 'gen1' to initiate rotation
+      rotateCsiKeysOlderThan: gen0 # user is not rotating CSI keys right now
+  # ...
+status:
+  cephx:
+    local:
+      keyGeneration: gen1 # because this is less than input 'gen2', local keys will be rotated
+      keyCephVersion: "19.2.0"
+    csi:
+      keyGeneration: gen0 # gen0 is the default assumption if gen is unknown
+      keyCephVersion: "" # unknown ceph version for key
+```
+
+I'm proposing `gen` as a prefix similar to `v` in semver versions to keep the same semantics when
+`rotateLocalKeysOlderThan: CurrentCephVersion` is configured. This "reads" better IMO, and it also
+guarantees that input values will be interpreted as a string and not confused with an int. We could
+easily use `v` as a prefix as well, instead of `gen`.
+
+This interface may be easier for users, but it does require reading the status to get the current
+`status.cephx.local.keyGeneration` then updating that. `DATETIME` inputs can use the value of `date`
+on the CLI or `time.Now()` in code without reading current status.
```

Comment:
Adding this for discussion per Travis's suggestion https://github.com/rook/rook/pull/15915#discussion_r2116425798